compiler/src/iree/compiler/Dialect/Flow/Transforms/CollapseDimensions.cpp

```diff
@@ -0,0 +1,232 @@
+    for (AffineMap map : genericOp.getIndexingMapsArray()) {
+      bool foundSeq = false;
+      for (auto [index, resultExpr] : llvm::enumerate(map.getResults())) {
+        if (resultExpr == nextExpr) {
+          foundSeq = (index > 0 && preExpr == map.getResult(index - 1));
+          break;
+        }
+      }
+      if (!foundSeq) return false;
+    }
+    return true;
+  };
+
+  ReassociationIndices range;
+  AffineExpr preExpr;
+  for (auto nextExpr : genericOp.getIndexingMapsArray().front().getResults()) {
+    unsigned pos = nextExpr.cast<AffineDimExpr>().getPosition();
+    if (!range.empty()) {
+      if (!hasAllMapsSameSequence(preExpr, nextExpr) || !pLoops.count(pos)) {
+        if (range.size() > 1)
+          contiguousLoops.push_back({range.begin(), range.end()});
+        range.clear();
+      }
+    }
+    preExpr = nextExpr;
+    if (pLoops.count(pos)) range.push_back(pos);
+  }
+  if (range.size() > 1) contiguousLoops.push_back(range);
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "Collapsing dimensions if possible: ";
+    for (auto indices : contiguousLoops) {
+      llvm::dbgs() << "[";
+      for (auto idx : indices) llvm::dbgs() << idx << ",";
+      llvm::dbgs() << "]\t";
+    }
+    llvm::dbgs() << "\n";
+  });
+
+  return contiguousLoops;
+}
+
+/// Collapse possible dimension of the given linalg.generic
+static FailureOr<SmallVector<Value>> collapseLinalgGeneric(
+    IRRewriter &rewriter, linalg::GenericOp genericOp,
+    SmallVector<ReassociationIndices> &collapseIndices) {
+  rewriter.setInsertionPoint(genericOp->getParentOp());
+  FailureOr<SmallVector<Value>> replacements =
+      mlir::linalg::collapseGenericOpIterationDims(genericOp, collapseIndices,
+                                                   rewriter);
+  if (failed(replacements) || replacements->empty()) {
+    return rewriter.notifyMatchFailure(genericOp,
+                                       "failed to collapse dimensions");
+  }
+
+  return replacements;
+}
+
+/// Returns true if the given op is collapsable.
+static bool isEligibleForCollapse(linalg::GenericOp genericOp) {
+  // TODO(guray) There is no mechanism to tell the collapsed indexes to
+  // `tensor.expand_shape`. Once we have this support in MLIR, we can enable
+  // dynamic tensor shapes.
+  if (genericOp.hasDynamicShape()) return false;
+
+  // TODO(guray) Currently we can only collapse when result of all the
+  // AffineMaps are dimensions. Possible to collapse cases like
+  // affine_map<d0, d1+d2> with affine_map<d0, d1+d2>, however, this is not
+  // supported in collapsing mechanism in MLIR. Once we have this support,
+  // we can remove this if statement.
+  if (llvm::any_of(genericOp.getIndexingMapsArray(), [](AffineMap map) {
+        return !map.isProjectedPermutation();
+      })) {
+    return false;
+  }
+
+  // IndexOp allows accesing induction variables. Collapsing might cause
+  // performance regression, so we disable it.
+  if (genericOp.hasIndexSemantics()) return false;
```

Comment:
Are we sure about this? I think the core transformation supports this. We should just drop this constraint....

Reply:
Yes you are right, the core transformation supports this. But it causes a performance regression in one CPU benchmark. I turned this off for a benchmark.

Here is the reproducer of slowdown. 
https://gist.github.com/grypp/b8c3f8fd6c2720b5a4f57c497e829812#file-collapse-mlir-L64-L67

I can look at the data access pattern instead of the hasIndexSemantics, but I am not sure what's the best way to do that.

Reply:
The collapsing shouldn’t happen when the loop index is used directly. So we should figure out why the upstream patterns allows it to happen. We don’t need to investigate the regression itself which is most likely due to wrong a miscompile.

Reply:
I dont see a reason why collapsing should happen. Upstream pattern handles it correctly, I dont see any problem with it.

Reply:
W.R.T regression, if it causes regression we can leave it off for now. Upstream it handles it for completeness. It is up to the user to either handle it or not.

Reply:
Let's file an issue and put the github issue ID to TODO here?

Reply:
As an alternative idea, I could create a flag and use it to turn off collapse in this regressed benchmark.